README.md
# bdsd - baos datapoint sdk daemon

This module intended to solve architecturing problem when developing js applcation for KNX.
Nodejs connects to BAOS 838 module via serial port, usually /dev/ttyAMA0 and it handles only one connection at a time.
But what if you want to connect to KNX bus multiple clients? 
If you want to write some custom scripts and use this device as a MQTT gateway at a same time?

So, schematically it may look like this:

![meow](./scheme.png)

# Installation

Assuming you have Raspberry Pi with installed Raspbian with configured access to BAOS module via /dev/ttyAMA0.

If not, follow instructions on [bobaos repository page](https://github.com/shabunin/bobaos#installation)

## Installation via script

```
$ curl -L https://raw.githubusercontent.com/bobaos/bdsd.sock/master/bdsd_install.sh | bash
```


## Manual installation

**1. Install npm package**

```
$ sudo npm install -g bdsd.sock --unsafe-perm
```

Check if it executes correctly:

```
$ bdsd.sock
Checking for leftover socket.
No leftover socket fount.
Listening at /run/user/1000/bdsd.sock
.....
.....
got bus state: connected
```

**2. Create systemd service folders, create service file**

```
$ cd ~/
$ mkdir ~/.config/systemd
$ mkdir ~/.config/systemd/user
$ touch ~/.config/systemd/user/bdsd.service
```

Then add following to this file using your favourite text editor:
```
[Unit]
Description=Bobaos Datapoint Sdk Daemon

[Service]
ExecStart=/usr/bin/env bdsd.sock

[Install]
WantedBy=default.target
```

**3. Enable service, enable automatic start-up**

```
$ systemctl --user daemon-reload
$ systemctl --user enable bdsd.service
$ sudo loginctl enable-linger pi
```

**4. Start the service**

```
$ systemctl --user start bdsd.service
```

 **5. Check with [bdsd-cli](https://github.com/shabunin/bdsd-cli)**

```
$ sudo npm -g install bdsd-cli
$ bdsd-cli
connected
bobaos> setProgrammingMode -v 1
Set programming mode: success
bobaos> getValue -s 1
{ id: 1, value: true } 
bobaos>
```
# Usage in js

For js applications there will be available client-side library which you may install from npm.

Further information you can find in repo: [bdsd.client](https://github.com/shabunin/bdsd.client)

# Usage with other programming languages

For other programming languages you may implement client-side library which should communicate with IPC following [Bobaos Datapoint Sdk Message Protocol](./PROTOCOL.md).


CHANGELOG
commit 35524bec5c08ba0881739b610c90f61cf358de26
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Sun Apr 8 18:06:39 2018 +0000

    Install script

commit 71d5d668ceb914b3a6627728ec636f8da00bab88
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Sun Apr 8 21:02:26 2018 +0300

    CHANGELOG

commit eeddb1ca5a321d1fcac0f26df72b6ce377970ef1
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Sun Apr 8 18:02:19 2018 +0000

    1.0.5

commit c5100e42c52add87c23e7841a472c35c46268e19
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Sun Apr 8 18:02:06 2018 +0000

    1.0.4

commit b9fe8e094bfbbadd6cc998c8d0e718fdb2e0fb98
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Sun Apr 8 18:01:55 2018 +0000

    Return raw buffer value on getValue/broadcast

commit 9190e038ac0daa42c09ed0e57715d84ad33bdb15
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Sun Apr 8 17:35:36 2018 +0000

    On 'set datapoint' at first get new value then broadcast to other clients

commit 13c2de3e3c873a998e4a98b805baff83bb5263ee
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Mon Mar 26 23:00:44 2018 +0300

    1.0.3

commit 9b01d224e93d3b689f6de67b3e2a8f5c52ae2ac6
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Mon Mar 26 23:00:35 2018 +0300

    Broadcast except sender on set value

commit be0c9004462f6437782efffcbbddd32a5015fae5
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Thu Mar 22 03:17:22 2018 +0300

    README

commit 8dd728d296185852669d985d4f47c1db93e9f744
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Thu Mar 22 03:13:38 2018 +0300

    preps to npm package

commit 6040317476d3a9b09b39695ed3f4aa93eb3b1314
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Tue Mar 13 16:20:31 2018 +0300

    README

commit c81b3bc13abec4f2f9d103e5d0910286a41ac6fe
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Tue Mar 13 16:19:17 2018 +0300

    README

commit cbbf2549ba13d6cd62c4b7c6d46da63f9e667137
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Tue Mar 13 16:09:45 2018 +0300

    PROTOCOL.md

commit b8292256b5892129090987917a0ca98556f5b564
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Tue Mar 13 12:57:40 2018 +0000

    Programming mode support

commit 6316c8d2d66bd41529c5546c5e3d3945f4c150ae
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Mon Mar 5 10:28:24 2018 +0000

    README

commit ab5f341176197eb7e40668e81ef37b885ede2ea9
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Thu Mar 1 16:18:41 2018 +0000

    correct service file

commit 5cb3662ca41a50997b0fea9e354f1cbaf06b7dd6
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Thu Mar 1 15:49:47 2018 +0000

    Installation instructions

commit 2875861857ac4c5d36a3500e8c6fceb70cabe482
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Thu Mar 1 15:42:34 2018 +0000

    README fixes

commit 0b5ed72cacd3d74634070d51e27a361785eb7ffd
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Thu Mar 1 15:42:02 2018 +0000

    README fixes

commit ebdee796721b93f791e2e9a790e585b82b550056
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Thu Mar 1 15:36:27 2018 +0000

    Systemd user service

commit e9d198f59d02a6c2de9686a3a10f2cf743a38238
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Thu Mar 1 15:32:28 2018 +0000

    Systemd user service

commit e5d754c81c4cef8acab00c78fcbb792ed333ccdb
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 20:33:25 2018 +0300

    typo

commit 6221a208a2fdbf8d6b13d933898f393a77c18a3b
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 19:46:41 2018 +0300

    README

commit e929524faea91ac8abc59301771639490ebe16d8
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 19:35:34 2018 +0300

    work in progress

commit 4d3f40535ae8dd7bc846c0b132720cea497198e0
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 19:30:00 2018 +0300

    work in progress

commit 43127deb5c4e321a852e2c2a9c11ff0bb532660c
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 14:28:08 2018 +0000

    1.0.2

commit d267c98caeb7d5c1ce7d7044bac68836c18db99e
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 14:27:42 2018 +0000

    Work in progress

commit 2becf4b8682bff26c9561c47df0d6d3ae45e775d
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 14:00:39 2018 +0000

    Reject when can't find datapoint

commit 6ed8a920557641cd29bd97b410fb16f864aa3133
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 15:36:47 2018 +0300

    1.0.1

commit 2d868a210eac04dab9662b2d208992195139d6af
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 15:36:39 2018 +0300

    README instructions

commit 30ddd663d9aca0d8affb548576ecd6818b21d3f2
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 15:35:56 2018 +0300

    switch/case default break

commit e3580fdb12cda117da74c1ef0f122c385c8f92f1
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Wed Feb 28 11:36:24 2018 +0000

    README in process

commit cfe65ce23e3677d21238ce808350b491f85dea79
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Sun Feb 25 04:19:44 2018 +0300

    start.sh

commit 843085a6cdbab1da781776b05b0031f8a983478c
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Sun Feb 25 00:49:33 2018 +0000

    work in progress

commit d8cb6a471fa5847c627fd28ccc9255abf0db111a
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Sun Feb 25 03:46:23 2018 +0300

    start.sh

commit c5cb1c752d68fcc8887dff9b092b5eb6e2d39812
Author: Vladimir Shabunin <va.shabunin@physics.msu.ru>
Date:   Sat Feb 24 23:45:56 2018 +0000

    README

commit 857a673b77e1624efd46acd291674ca9ee8f870b

datapointSdk/sdk.js
// Bobaos Datapoint Sdk
// collects datapoints information, provide methods to set, get, read values.
const DPTs = require('knx-dpts-baos');
const Baos = require('bobaos');
const EE = require('events').EventEmitter;

const Sdk = (params) => {
  let self = new EE();
  // simple storage for datapoint objects and descriptions
  self.store = {
    datapoints: [],
    descriptions: []
  };

  // serialport params
  let serialPortDevice = '/dev/ttyAMA0';
  let serialPortParams = {
    baudRate: 19200,
    parity: "even",
    dataBits: 8,
    stopBits: 1
  };
  if (typeof params !== 'undefined') {
    if (params.serialPort !== null && typeof params.serialPort === 'object') {
      if (params.serialPort.device !== null && typeof params.serialPort.device === 'string') {
        serialPortDevice = params.serialPort.device;
      }
      if (params.serialPort.params !== null && typeof params.serialPort.params === 'object') {
        serialPortParams = params.serialPort.params;
      }
    }
  }
  // init bobaos
  const bobaos = new Baos({serialPort: {device: serialPortDevice, params: serialPortParams}, debug: false});

  // register events

  // Datapoint class
  const Datapoint = function (props) {
    this.id = props.id;
    this.dpt = props.dpt;
    this.flags = props.flags;
    this.length = props.length;
    this.value = null;
  };
  Datapoint.prototype.getDescription = function() {
    return new Promise((resolve, reject) => {
      resolve({
        id: this.id,
        dpt: this.dpt,
        flags: this.flags,
        length: this.length
      });
    });
  };
  Datapoint.prototype.setValue = function (value) {
    return new Promise((resolve, reject) => {
      let id = this.id;
      let dpt = this.dpt;
      try {
        let encodedValue = DPTs[dpt].fromJS(value);
        bobaos.setDatapointValue(id, encodedValue)
          .then(resolve)
          .catch(reject);
      } catch (e) {
        reject(e);
      }

    });
  };
  Datapoint.prototype.getValue = function () {
    // TODO: refactor
    const processValuePayload = t => {
      return new Promise((resolve, reject) => {
        let id = t.id;
        let encodedValue = t.value;
        self.findDatapoint(id)
          .then(datapoint => {
            return datapoint
              ._applyValue(encodedValue);
          })
          .then(payload => {
            resolve(payload);
          })
          .catch(e => {
            reject(e)
          });
      });
    };

    let id = this.id;
    return bobaos
      .getDatapointValue(id, 1)
      .then(payload => {
        return processValuePayload(payload[0]);
      })
  };
  Datapoint.prototype.readFromBus = function () {
    return new Promise((resolve, reject) => {
      let id = this.id;
      let length = this.length;
      bobaos
        .readDatapointFromBus(id, length)
        .then(payload => {
          resolve(payload);
        })
        .catch(e => {
          reject(e);
        });
    });
  };
  // to internal use. when we got value from bus put it to store
  Datapoint.prototype._applyValue = function (value) {
    return new Promise((resolve, reject) => {
      let id = this.id;
      let dpt = this.dpt;
      try {
        this.value = DPTs[dpt].toJS(value);
        resolve({value: this.value, raw: value});
      } catch (e) {
        reject(e);
      }
    });
  };
  // function to get all datapoints descriptions
  self.getAllDatapointDescriptions = _ => {
    return new Promise((resolve, reject) => {
      try {
        resolve(self.store.descriptions.slice());
      } catch (e) {
        reject(e);
      }
    });
  };

  // function to get datapoint object. Example: .datapoint(1).getValue();
  self.findDatapoint = id => {
    return new Promise((resolve, reject) => {
      const findDatapointById = t => t.id === id;
      let datapointIndex = self.store.datapoints.findIndex(findDatapointById);
      if (datapointIndex >= 0) {
        resolve(self.store.datapoints[datapointIndex]);
      } else {
        // cannot find datapoint with this id
        reject(new Error('cannot find datapoint with id:' + id));
      }
    });
  };


  // programming mode
  self.setProgrammingMode = value => {
    return new Promise((resolve, reject) => {
      const valueBuff = Buffer.alloc(1, value & 0x01);
      bobaos
        .setServerItem(15, valueBuff)
        .then(payload => {
          resolve(payload);
        })
        .catch(e => {
          reject(e);
        })
    });
  };


  // 1. set server item for indication to false at beginning
  // 2. get description for all datapoints [1-1000].
  // 3. send GetServerItem request for "bus connected state" item.
  // 4. enable indications
  // enable/disable indications steps are done to be
  // sure that we have all datapoint descr when got ind event
  const setIndications = function (state) {
    let value = state ? 1 : 0;
    bobaos.setServerItem(17, Buffer.alloc(1, value))
      .then(_ => {
        console.log('success on setting indications to:', value);
      })
      .catch(e => {
        console.log('error while setting indications to:', value, e);
      })
  };
  const getAllDatapointDescription = _ => {
    const processDatapointDescription = payload => {
      if (Array.isArray(payload)) {
        payload.forEach(t => {
          let datapoint = new Datapoint(t);
          console.log('success on get datapoint description: { id:', datapoint.id, ', dpt: ', datapoint.dpt, '}');
          self.store.datapoints.push(datapoint);
          self.store.descriptions.push(t);
        });
      }
    };
    const processError = e => {
      //console.log('error while getting datapoint description', e);
    };
    // clear store
    self.store.datapoints = [];
    self.store.descriptions = [];
    // how much datapoints at one request
    const number = 30;
    for (let i = 0, imax = 1000; i < imax; i += number) {
      if ((imax - i) > number) {
        bobaos.getDatapointDescription(i + 1, number)
          .then(processDatapointDescription)
          .catch(processError);
      } else {
        bobaos.getDatapointDescription(i + 1, imax - i)
          .then(processDatapointDescription)
          .catch(processError);
      }
    }
  };
  const getBusState = _ => {
    bobaos.getServerItem(10)
      .then(payload => {
        if (Array.isArray(payload)) {
          payload.forEach(t => {
            if (t.id === 10 && t.value.readUInt8(0) === 1) {
              console.log('got bus state: connected');
              self.emit('connected');
            }
          })
        }
      })
      .catch(e => {
        console.log('error while getting bus state', e);
      });
  };
  bobaos.on('open', _ => {
    console.log('connected to baos');
    // get all descriptions and after that get bus state
    setIndications(false);
    getAllDatapointDescription();
    setIndications(true);
    getBusState();
  });
  bobaos.on('reset', function () {
    console.log('got reset ind');
    // on reset indication. e.g. when you downloaded new config from ETS
    // get all descriptions and after that get bus state
    setIndications(false);
    getAllDatapointDescription();
    setIndications(true);
    getBusState();
  });
  // now process value indications
  bobaos.on('DatapointValue.Ind', payload => {
    const processValuePayload = function (t) {
      let id = t.id;
      let encodedValue = t.value;
      self.findDatapoint(id)
        .then(datapoint => {
          return datapoint
            ._applyValue(encodedValue);
        })
        .then(payload => {
          self.emit('DatapointValue.Ind', {id: id, value: payload.value, raw: payload.raw});
        })
        .catch(e => {
          // should never be executed but anyway
          console.log('error on DatapointValue.Ind', e);
        });
    };
    if (Array.isArray(payload)) {
      payload.forEach(t => {
        processValuePayload(t);
      })
    }
  });
  // handle bus indication events
  bobaos.on('ServerItem.Ind', payload => {
    if (Array.isArray(payload)) {
      payload.forEach(t => {
        if (t.id === 10) {
          if (t.value.readUInt8(0) === 1) {
            self.emit('bus connected');
          } else {
            self.emit('bus disconnected');
          }
        }
      })
    }
  });
  return self;
};

module.exports = Sdk;